src/quic/streams.cc

```diff
@@ -1,61 +1,1049 @@
+    Stream* stream;
+    ASSIGN_OR_RETURN_UNWRAP(&stream, args.Holder());
+    BaseObjectPtr<Blob::Reader> reader = stream->get_reader();
+    if (reader) return args.GetReturnValue().Set(reader->object());
+    THROW_ERR_INVALID_STATE(Environment::GetCurrent(args),
+                            "Unable to get a reader for the stream");
+  }
+};
+
+// ============================================================================
+
+class Stream::Outbound final : public MemoryRetainer {
+ public:
+  Outbound(Stream* stream, std::shared_ptr<DataQueue> queue)
+      : stream_(stream),
+        queue_(std::move(queue)),
+        reader_(queue_->get_reader()) {}
+
+  void Acknowledge(size_t amount) {
+    size_t remaining = std::min(amount, total_ - uncommitted_);
+    while (remaining > 0 && head_ != nullptr) {
+      DCHECK_LE(head_->ack_offset, head_->offset);
+      // The amount to acknowledge in this chunk is the lesser of the total
+      // amount remaining to acknowledge or the total remaining unacknowledged
+      // bytes in the chunk.
+      size_t amount_to_ack =
+          std::min(remaining, head_->offset - head_->ack_offset);
+      // If the amount to ack is zero here, it means our ack offset has caught
+      // up to our commit offset, which means there's nothing left to
+      // acknowledge yet. We could treat this as an error but let's just stop
+      // here.
+      if (amount_to_ack == 0) break;
+
+      // Adjust our remaining down and our ack_offset up...
+      remaining -= amount_to_ack;
+      head_->ack_offset += amount_to_ack;
+
+      // If we've fully acknowledged this chunk, free it and decrement total.
+      if (head_->ack_offset == head_->buf.len) {
+        DCHECK_GE(total_, head_->buf.len);
+        total_ -= head_->buf.len;
+        // if tail_ == head_ here, it means we've fully acknowledged our current
+        // buffer. Set tail to nullptr since we're freeing it here.
+        if (head_.get() == tail_) {
+          // In this case, commit_head_ should have already been set to nullptr.
+          // Because we should only have hit this case if the entire buffer
+          // had been committed.
+          DCHECK(commit_head_ == nullptr);
+          tail_ = nullptr;
+        }
+        head_ = std::move(head_->next);
+        DCHECK_IMPLIES(head_ == nullptr, tail_ == nullptr);
+      }
+    }
+  }
+
+  void Commit(size_t amount) {
+    // Commit amount number of bytes from the current uncommitted
+    // byte queue. Importantly, this does not remove the bytes
+    // from the byte queue.
+    size_t remaining = std::min(uncommitted_, amount);
+    // There's nothing to commit.
+    while (remaining > 0 && commit_head_ != nullptr) {
+      // The amount to commit is the lesser of the total amount remaining to
+      // commit and the remaining uncommitted bytes in this chunk.
+      size_t amount_to_commit =
+          std::min(remaining, commit_head_->buf.len - commit_head_->offset);
```

Comment:
Seem to be getting a conflicting types error from this: https://github.com/nodejs/node/actions/runs/5115751061/jobs/9197295402?pr=48244#step:6:3002